sections/design/basics-of-react.tex

```diff
@@ -0,0 +1,97 @@
+In the function we return JSX which React will then compile to html.
+The idea is to build the UI from components by exporting them and then importing them into other larger components to contain them. This means that the components act as the building blocks of the UI because we can reuse them in other components.
+\begin{lstlisting}
+import React from "react";
+        
+function Button() {
+    return (
+        <div>
+            <button onClick={console.log('Hello')}>Hello React</button>
+        </div>
+    );
+}
+    
+export default Button
+\end{lstlisting}
+As an example we can create a button component with an on-click event embedded into it. 
+Here we use brackets to enable us to embed JavaScript in the JSX which will be run when the button is clicked.
+We can now export this button and it can be imported by other components. 
+A component is conceptually a JavaScript function, which means that they can accept inputs called \texttt{props} which can contain data that can be used in the react elements. If a component is used in another component, we can use \texttt{props} to pass data to the child component.
+\\\\
+In general, \texttt{props} should be read-only, which means a component should try to change it's values. 
+To allow the UI elements to change over time we instead make use of \textit{State} which allows the components to change their output over time when handling user inputs or network responses. 
+% Skriv om stateful og stateless components
+\textit{State}  is similar to \texttt{props} but is instead private and only the specific component can make changes to it. To make use of \textit{State} we need to initialize the component as a class instead of a function
+\begin{lstlisting}
+    function Button() {
+        return (
+            <div>
+                <button onClick={console.log('Hello')}>Hello React</button>
+            </div>
+        );
+    }
+\end{lstlisting}
+We can convert this into a class like this:
+\begin{lstlisting}
+    class Button extends React.Component {
+        render() {
+            return (
+                <div>
+                    <button onClick={console.log('Hello')}>Hello React</button>
+                </div>
+            );
+        }
+    }
+\end{lstlisting}
+This is useful since the render method is called everytime the local \textit{State} is changed. We can set the intial state of a component by assigning to \texttt{this.state}
+\begin{center}
+    \texttt{this.state = \{date = new Date();\};}
+\end{center}
+The state can then be changed by reassigning to the \texttt{date} variable. This is done by using the \texttt{this.setState()} function.
+\begin{center}
+    \texttt{this.setState(\{date = new Date();\});}
+\end{center}
+When a component has a state set it is considered to be a stateful component. 
+Parent or child components should not care if a certain component is stateful or stateless since the state is local and not accessible to other components. 
```

Comment:
```suggestion
Parent or child components should not care if a certain component is stateful or stateless since the state is local and not accessible by other components. 
```